include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td

```diff
@@ -282,15 +285,28 @@ def AndOfPad : Pat <
   (MoveNameHint $old, (PadPrimOp (AndPrimOp (TailPrimOp $x, (TypeWidthAdjust32 $x, $y)), $y), $n)),
   [(KnownWidth $x), (UIntType $x), (EqualTypes $x, $pad)]>;
 
-/// cvt is adding a 0, which will pass through the and
-/// and(cvt(x:U), const) -> cat(0, and(x,trunc(const))
+def AndOfAsSIntL : Pat<
+  (AndPrimOp:$old (AsSIntPrimOp $x), $y),
+  (MoveNameHint $old, (AndPrimOp $x, (AsUIntPrimOp $y))),
+  []>;
+
+def AndOfAsSIntR : Pat<
+  (AndPrimOp:$old $x, (AsSIntPrimOp $y)),
+  (MoveNameHint $old, (AndPrimOp (AsUIntPrimOp $x), $y)),
+  []>;
+
+
+/// Width narrowing
+/// Only trigger when cat is full length
+/// and(cat(0, x:U), const) -> cat(0, and(x,zextortrunc(const))
 def AndCvtU : Pat <
```

Comment:
Nit: the name `AndCvtU` could be updated to better reflect what the pattern now literally is.